framework/src/userobject/NodalNormalsUserObject.C

```diff
@@ -0,0 +1,181 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "NodalNormalsUserObject.h"
+#include "MooseMesh.h"
+
+template <>
+InputParameters
+validParams<NodalNormalsUserObject>()
+{
+  InputParameters params = validParams<NodalUserObject>();
+  return params;
+}
+
+Threads::spin_mutex NodalNormalsUserObject::_add_contr;
+
+NodalNormalsUserObject::NodalNormalsUserObject(const InputParameters & parameters)
+  : NodalUserObject(parameters),
+    _nodal_normals(declareRestartableData<std::map<dof_id_type, Point>>("nodal_normals"))
+{
+}
+
+void
+NodalNormalsUserObject::execute()
+{
+}
+
+void
+NodalNormalsUserObject::initialize()
+{
+}
+
+void
+NodalNormalsUserObject::finalize()
+{
+}
+
+void
+NodalNormalsUserObject::threadJoin(const UserObject & uo)
+{
+  const NodalNormalsUserObject & nnuo = dynamic_cast<const NodalNormalsUserObject &>(uo);
+
+  for (auto & it : nnuo._nodal_normals)
+    add(it.first, it.second);
+}
+
+Point
+NodalNormalsUserObject::getNormal(dof_id_type node_id) const
+{
+  const auto & it = _nodal_normals.find(node_id);
+  if (it != _nodal_normals.end())
+    return it->second;
+  else
+    return Point(0, 0, 0);
+}
+
+void
+NodalNormalsUserObject::add(const Node * node, RealGradient grad) const
+{
+  Threads::spin_mutex::scoped_lock lock(_add_contr);
```

Comment:
Do we really need the lock here? Seems like you don't need it until the other `add()` method.

Reply:
This is called from `NodalNormalsBoundaryNodes`. The other `add` is also called from `deserialization`. So may be just move this lock inside the if-statement below.

Reply:
I thought about this more and I do not think we want to move this. This is called from different threads and if we moved the lock, we would have unprotected access into `_mesh`.